### PR TITLE
make analysis of session a separate background task

### PR DESCRIPTION
**Pull request by mzfr**
This PR will separate the analysis and deletion of the session.

### CODE REVIEW

File: tanner/server.py [hunk 1]
```diff
@@ -112,11 +114,20 @@ async def on_shutdown(self, app):
...
+                    self.redis_client, self.pg_client
+                )
                 await asyncio.sleep(self.delete_timeout)
         except asyncio.CancelledError:
             pass
 
+    async def analyze_sessions(self):
+        # try:
+        while True:
+            await self.session_analyzer(self.redis_client, self.pg_client)
```
Comment (afeena) [inline tanner/server.py:127]:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/server.py", line 127, in analyze_sessions
    await self.session_analyzer(self.redis_client, self.pg_client)
AttributeError: 'TannerServer' object has no attribute 'session_analyzer'
```

Comment (mzfr) [reply tanner/server.py:127]: Yes I found that issue.

**Comment by afeena** [comment]
don't forget to cancel the task in `cleanup_background_tasks`

File: tanner/sessions/session_analyzer.py [hunk 2]
```diff
@@ -18,18 +18,20 @@ def __init__(self, loop=None):
...
         else:
-            result = await self.create_stats(session, redis_client)
-            await self.queue.put(result)
-            await self.save_session(redis_client, pg_client)
+            for key in keys:
+                session = await redis_client.get(key, encoding='utf-8')
+                session = json.loads(session)
+                result = await self.create_stats(session, redis_client)
+                await self.queue.put(result)
+                await self.save_session(redis_client, pg_client)
```
Comment (afeena) [inline tanner/sessions/session_analyzer.py:34]: do we really need a separate save function or we can do everything in current function?

anyway, save should not be called after you put each session into the queue, it should be called once

File: tanner/sessions/session_manager.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ async def delete_session(self, sess, redis_client, pg_client):
             await sess.remove_associated_env()
         try:
             await redis_client.set(sess.get_uuid(), sess.to_json())
-            await self.analyzer.analyze(sess.get_uuid(), redis_client, pg_client)
+            # await self.analyzer.analyze(sess.get_uuid(), redis_client, pg_client)
```
Comment (afeena) [inline tanner/sessions/session_manager.py:98]: remove commented line

File: tanner/sessions/session_analyzer.py [hunk 4]
```diff
@@ -18,18 +18,20 @@ def __init__(self, loop=None):
...
 
     async def save_session(self, redis_client, pg_client):
         while not self.queue.empty():
```
Comment (afeena) [inline tanner/sessions/session_analyzer.py:37]: you have try for redis error, but not postgres errors

File: tanner/sessions/session_analyzer.py [hunk 5]
```diff
@@ -15,71 +16,90 @@ class SessionAnalyzer:
...
+                        await redis_client.delete(*[del_key])
+                    except psycopg2.ProgrammingError as pg_error:
+                        self.logger.exception(
+                            "Error with Postgres %s. Session with session-id %s will not be deleted",
+                            pg_error,
+                            key,
+                        )
+                    except aioredis.ProtocolError as redis_error:
+                        self.logger.exception(
+                            "Error with redis. Session will be returned to the queue: %s",
```
Comment (afeena) [inline tanner/sessions/session_analyzer.py:61]: there is no queue anymore to return the session to

File: tanner/sessions/session_analyzer.py [hunk 6]
```diff
@@ -15,71 +16,90 @@ class SessionAnalyzer:
...
+                            redis_error,
+                        )
+                # This is the key which stores all the dorks.
+                # It matches the pattern of other keys.
+                except aioredis.errors.ReplyError:
```
Comment (afeena) [inline tanner/sessions/session_analyzer.py:66]: can we avoid this by using another pattern for sessions?

**Comment by afeena** [comment]
@mzfr could you please fix the tests?

**Comment by mzfr** [comment]
@afeena Yes I am working on them, Trying to understnad why they are failing.

**Pull request merged**